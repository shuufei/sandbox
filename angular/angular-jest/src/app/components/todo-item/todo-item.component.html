<h1 class="title">{{title}}</h1>
<span class="completed">{{isCompleted ? 'completed' : 'todo'}}</span>
<button class="toggle-button" (click)="toggle()">toggle!</button>