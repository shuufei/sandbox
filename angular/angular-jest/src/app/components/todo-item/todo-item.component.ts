import { ChangeDetectionStrategy, Component, Input, OnInit } from '@angular/core';

type LocalState = {
  title: string;
  isCompleted: boolean;
};
@Component({
  selector: 'app-todo-item',
  templateUrl: './todo-item.component.html',
  styleUrls: ['./todo-item.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TodoItemComponent implements OnInit {
  @Input()
  set title(value: LocalState['title']) {
    this.state.title = value;
  }
  get title(): LocalState['title'] {
    return this.state.title;
  }

  @Input()
  set isCompleted(value: LocalState['isCompleted']) {
    this.state.isCompleted = value;
  }
  get isCompleted(): LocalState['isCompleted'] {
    return this.state.isCompleted;
  }

  private state: LocalState = {
    title: '',
    isCompleted: false
  };

  constructor() { }

  ngOnInit(): void {
  }

  toggle(): void {
    this.state.isCompleted = !this.state.isCompleted;
  }

}
