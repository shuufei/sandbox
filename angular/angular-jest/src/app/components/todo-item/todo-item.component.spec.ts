import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { runChangeDetection } from 'src/tests/changeDetectionForOnPush';

import { TodoItemComponent } from './todo-item.component';

describe('TodoItemComponent', () => {
  let component: TodoItemComponent;
  let fixture: ComponentFixture<TodoItemComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ TodoItemComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TodoItemComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('コンポーネントに指定したタイトルが表示される', async () => {
    const title = 'hanashiro';
    component.title = title;
    await runChangeDetection(fixture);
    expect(fixture).toMatchSnapshot();
  });

  test(`コンポーネントのisCompletedにfalseを指定したとき、状態表示部分に 'todo' が表示される`, async () => {
    component.isCompleted = false;
    await runChangeDetection(fixture);
    expect(fixture).toMatchSnapshot();
  });

  test(`コンポーネントのisCompletedにtrueを指定したとき、状態表示部分に 'completed' が表示される`, async () => {
    component.isCompleted = true;
    await runChangeDetection(fixture);
    expect(fixture).toMatchSnapshot();
  });

  test(`isCompletedがtrueの状態でtoggleボタンを押下したとき、状態表示部分に 'todo' が表示される`, async () => {
    component.isCompleted = true;
    await runChangeDetection(fixture);
    const buttonDebugEl = fixture.debugElement.query(By.css('.toggle-button'));
    (buttonDebugEl.nativeElement as HTMLButtonElement).click();
    await runChangeDetection(fixture);
    expect(fixture).toMatchSnapshot();
  });

  test(`isCompletedがfalseの状態でtoggleボタンを押下したとき、状態表示部分に 'completed' が表示される`, async () => {
    component.isCompleted = false;
    await runChangeDetection(fixture);
    const buttonDebugEl = fixture.debugElement.query(By.css('.toggle-button'));
    (buttonDebugEl.nativeElement as HTMLButtonElement).click();
    await runChangeDetection(fixture);
    expect(fixture).toMatchSnapshot();
  });

});
