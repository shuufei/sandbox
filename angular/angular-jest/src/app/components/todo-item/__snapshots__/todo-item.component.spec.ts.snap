// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItemComponent isCompletedがfalseの状態でtoggleボタンを押下したとき、状態表示部分に 'completed' が表示される 1`] = `
<app-todo-item
  state={[Function Object]}
>
  <h1
    class="title"
  >
    
  </h1><span
    class="completed"
  >
    completed
  </span><button
    class="toggle-button"
  >
    toggle!
  </button>
</app-todo-item>
`;

exports[`TodoItemComponent isCompletedがtrueの状態でtoggleボタンを押下したとき、状態表示部分に 'todo' が表示される 1`] = `
<app-todo-item
  state={[Function Object]}
>
  <h1
    class="title"
  >
    
  </h1><span
    class="completed"
  >
    todo
  </span><button
    class="toggle-button"
  >
    toggle!
  </button>
</app-todo-item>
`;

exports[`TodoItemComponent コンポーネントに指定したタイトルが表示される 1`] = `
<app-todo-item
  state={[Function Object]}
>
  <h1
    class="title"
  >
    hanashiro
  </h1><span
    class="completed"
  >
    todo
  </span><button
    class="toggle-button"
  >
    toggle!
  </button>
</app-todo-item>
`;

exports[`TodoItemComponent コンポーネントのisCompletedにfalseを指定したとき、状態表示部分に 'todo' が表示される 1`] = `
<app-todo-item
  state={[Function Object]}
>
  <h1
    class="title"
  >
    
  </h1><span
    class="completed"
  >
    todo
  </span><button
    class="toggle-button"
  >
    toggle!
  </button>
</app-todo-item>
`;

exports[`TodoItemComponent コンポーネントのisCompletedにtrueを指定したとき、状態表示部分に 'completed' が表示される 1`] = `
<app-todo-item
  state={[Function Object]}
>
  <h1
    class="title"
  >
    
  </h1><span
    class="completed"
  >
    completed
  </span><button
    class="toggle-button"
  >
    toggle!
  </button>
</app-todo-item>
`;
